
# Preparations

1. These scripts only work when the the directory structure for your work looks like:  
       working_directory/  
                  scripts/ (All CoCena? scripts)  
                  data/ (count table)  
                  reference_files/ (gmt file, TFcat file etc)  
                  sample_info/ (metadata file)  

2. The following parameters have to be set:

  + **working_directory**:  
     Provide path for the working directory containing the following folders: 
     scripts, data, reference files, samples info, ensure you **end the path with a forward slash**
   
  + **count_file_name**:  
     Input the name of the count file in csv format (*path: working_directory/data/count_file.csv*)  
     The file structure should be the following:  
      + blank,sample1,sample2,sample3,sampleN    
      + gene1, count, count, ...  
      + gene2, count, count, ...  
      + geneN, count, count, ...
    
  + **topvar_genes**:  
     The number of genes to be included for the co-expression analysis based on their variance, 
     e.g. if topvar_genes=500, the genes will be sorted by their variance (high to low) and the first 500 genes
     (genes which show most variance) will be selected for the downstream analysis.  
     If you do not know what to input, leave the options as it is. topvar_genes depends on the data, 
     e.g. it is different for single-cell and bulk data.

  + **info_dataset_name**:  
     Input the name of the metadata file in csv format (*path: working_directory/sample_info/info_Dataset*).
     The file provides metadata of samples, samples will be grouped by variables defined in this file.  
     The file structure should be the following:  
       + file (rownames = sample names (which correspond to the first column of the count file))  
       + first colname empty, Condition, Disease, Tissue......    
       + GSM111456 Control Influenza PBMC ...
        
  + **voi** - Variable of interest:  
     Name of the column in the info data set which contains the grouping information for calculating 
     group fold changes (GFCs), e.g. if one is interested in disease then the voi needs to be set to the 
     column name containing disease information in the info data set.
     
  + **TF_list_name**:  
     Input the name of the transcription factor file (*path: working_directory/reference_files/TF_file*).
     The file structure should be the following:
      + file column names Taxon1 Taxon2 Type_of_gene
      + Mouse Human Merged_Taxa
      + Aatf AATF TF
      + Aes AES Co_factor  

  + **GMT files**:  
     GMT: Gene Matrix Transposed file format (.gmt), see
     [link](https://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats) for details.
     Input the name of the representative GMT files (*path: working_directory/reference_files/gmtfile*). 
     Only the gmtfile_name_hallmarks is essential for the ClusterProfiler GSEA. Others can be provided for
     coloring of the network nodes.  

  + **organism**:
    Input name of the study organism human or mouse.

  + **min_corr** and **range_cutoff_length**:  
     To select genes to be used for the subsequent analysis, the optimal correlation coefficient cut-off needs
     to be determined. 
     The cutoff range to be tested is defined by min. correlation limit specified below, up to the max.
     correlation found in data between genes, while the number of cut-offs tested is determined by
     range_cutoff_length. E.g. if min_corr=0.5 and the range_cutoff_length=300, 300 cutoff-values between
     0.5-max(correlation_between_genes) will be tested. For bulk RNA data, 0.5 gives decent results might need
     to lower this for single cell experiments.  
     
  + **print_distribution_plots**:  
     True will print the distribution plots for all tested cut-offs to pdf files.  
     
  + **min_nodes_number_for_network** and **min_nodes_number_for_cluster**:  
     Input the minimal number of nodes to determine a separate network (=igraph component). For each network
     a layout is calculated, which are then merged to obtain one network plot. the minimal number of nodes to
     determine a cluster for the calculation of clusters by the cluster algorithm. 

  + **data_in_log**:  
     Is data on logarithmic scale?  
     
  + **range_GFC**:  
    Input the range the group fold change (GFC) should have.  
    
  + **layout_algorithm**:
     Input the layout algorithm to be used for the network layout 
     (choose between layout_with_fr = fast or cytoscape = slow, but nicer separation of the clusters calculated.      *cytoscape needs to be installed and open!*)
     
```{r}
# object_keep <- c("norm_anno", "sample_table")
# rm(list=setdiff(ls(), object_keep))
# gc()
```

     

```{r}
working_directory = paste0(getwd(), "/", sep="")

count_file_name <- norm_anno
count_file_name <- count_file_name[,c(2:(ncol(count_file_name)-3))]
count_file_name_symbol <- count_file_name[!duplicated(count_file_name[,ncol(count_file_name)]), ]
count_file_name_symbol <- count_file_name_symbol[complete.cases(count_file_name_symbol), ]

row.names(count_file_name_symbol) <- count_file_name_symbol[,ncol(count_file_name)]
count_file_name <- count_file_name_symbol[,c(-(ncol(count_file_name)))]


topvar_genes = 10000


info_dataset_name <- sample_table
rownames(info_dataset_name)<-info_dataset_name$ID

voi = "merged"
info_dataset_name$ID <- row.names(info_dataset_name)


TF_list_name = "TFcat.txt"
gmtfile_name_hallmarks = "h.all.v6.1.symbols.gmt"
gmtfile_name_go = "c5.bp.v7.0.symbols.gmt"
gmtfile_name_kegg = "c2.cp.kegg.v7.0.symbols.gmt"
gmtfile_name_reactome = "c2.cp.reactome.v7.0.symbols.gmt"

organism = "human"

min_corr=0.7
range_cutoff_length=1000
print_distribution_plots = FALSE

min_nodes_number_for_network=40 
min_nodes_number_for_cluster=40

data_in_log=F

range_GFC=2.0

layout_algorithm = "layout_with_fr"  


```



# Packages

### Install CRAN
```{r}
# CRAN packages
list.of.packages_CoCena <- c("bench",
                   "BiocManager",
                   "combinat",
                   "ComplexHeatmap",
                   "dplyr",
                   "ggnetwork",
                   "ggplot2",
                   "graphics",
                   "grDevices",
                   "grid",
                   "gtools",
                   "Hmisc",
                   "igraph",
                   "intergraph",
                   "knitr",
                   "MCDA",
                   "openxlsx",
                   "pals",
                   "pheatmap",
                   "purrr",
                   "stringi",
                   "tidyr",
                   "tidyverse",
                   "utils")

new.packages_CoCena <- list.of.packages_CoCena[!(list.of.packages_CoCena %in% installed.packages()[,"Package"])]

if(length(new.packages_CoCena)>0) install.packages(new.packages_CoCena)
```

### Install BioConductor

```{r}
# BioconductoR packages
list.of.bioc.packages_CoCena<- c("biomaRt",
                       "clusterProfiler",
                       "DOSE",
                       "org.Hs.eg.db",
                       "pcaGoPromoter.Hs.hg19",
                       "RCy3",
                       "ReactomePA")

new.packages.bioc_CoCena <- list.of.bioc.packages_CoCena[!(list.of.bioc.packages_CoCena %in% installed.packages()[,"Package"])]
 
if(length(new.packages.bioc_CoCena)>0)if (!requireNamespace("BiocManager")) install.packages("BiocManager")
BiocManager::install(new.packages.bioc_CoCena, update = FALSE)
```

### Load Packages

```{r load packages, results='hide',message=FALSE,warning=FALSE}
invisible(lapply(c(list.of.packages_CoCena,new.packages.bioc_CoCena), require, character.only = TRUE))
```

# Data import

```{r}
mart <- biomaRt::useMart("ensembl")
mart <- biomaRt::listDatasets(mart) 
human <- biomaRt::useMart("ensembl", dataset = "hsapiens_gene_ensembl") 
mouse <- biomaRt::useMart("ensembl", dataset = "mmusculus_gene_ensembl") 


count_table <- count_file_name

universe_Entrez <- clusterProfiler:: bitr(row.names(count_table), 
                                         fromType="SYMBOL", 
                                         toType="ENTREZID", 
                                         OrgDb="org.Hs.eg.db", 
                                         drop = T)


info_dataset <- info_dataset_name


TF_list <- read.delim(paste0(working_directory, "reference_files/", TF_list_name),
                      header=TRUE,
                      check.names=F)

gmtfile_hallmarks <- clusterProfiler::read.gmt(paste0(working_directory, "reference_files/", gmtfile_name_hallmarks))
gmtfile_go <- clusterProfiler::read.gmt(paste0(working_directory, "reference_files/", gmtfile_name_go))
gmtfile_kegg <- clusterProfiler::read.gmt(paste0(working_directory, "reference_files/", gmtfile_name_kegg))
gmtfile_reactome <- clusterProfiler::read.gmt(paste0(working_directory, "reference_files/", gmtfile_name_reactome))
```



# Analysis

## Data filtering based on variability

Filtering data based on the intra-gene variance, will select first `r topvar_genes` genes based on their variance.

```{r}
ds = count_table[order(apply(count_table,1,var), decreasing=T),]
dd2 <- head(ds,topvar_genes)
dd2 = t(dd2)
```


## Correlation and correlation cut-off calculation

The first script *correlation_actions.R* calculates the **correlation_df** containing the information about the correlation of each gene to all other genes with respective p-value and r-value. Moreover, the script corrects for multiple testing, and filters the correlation_df for pval < 0.05 and r values > 0 resulting in **correlation_df_filt**. Additionally, the script calculates the cut-off values to be tested (**range_cutoff**).  

The second script *obtain_cutoff_stats.R* contains the function **cutoff_prep** for obtaining stats for each cutoff, namely R?, the number of edges (no_edges), the number of nodes (no_nodes), as well as the number of networks (no_of_networks) the resulting igraph object would have.   
In addition, all degree distribution plots are saved in pdf format in a newly generated directory (if print_distribution_plots = T in *Preparations*). 

The third script *optimal_cutoff.R* calculates the optimal cut-off.

```{r}
source(paste0(working_directory,"scripts/", "correlation_actions.R"))

source(paste0(working_directory,"scripts/", "obtain_cutoff_stats.R"))

cutoff_stats = do.call("rbind", lapply(X = range_cutoff,
                                       FUN = cutoff_prep,
                                       corrdf_r = correlation_df_filt,
                                       print.all.plots = print_distribution_plots))

source(paste0(working_directory,"scripts/", "optimal_cutoff.R"))
```


Execute the following chunk to see the table containing all calculated cut-offs as well as to see the degree distribution plot for the calculated optimal cut-off.

```{r, results='asis'}
kable(cutoff_stats_concise, caption = "Correlation cut-off stats")
plot(dd_plot_calculated_optimal)

```


## Data filtering based on correlation cut off

**Set the optimal_cutoff your want to use for the further analysis.**  
A new directory for the chosen optimal_cutoff is generated in which all further results will be saved.   
The degree distribution plot for the chosen correlation cut-off is plotted and saved in the generated directory. In addition the count data is filtered (**filt_cutoff_counts**) to maintain only genes with at least one correlation to another gene above the chosen correlation cut-off. These genes are represented within a heatmap of the filtered count data.


```{r}


optimal_cutoff = 0.857 #*********

#optimal_cutoff = calculated_optimal_cutoff


cutoff_wd <- paste0("dir_",optimal_cutoff, "_", topvar_genes)
if(!cutoff_wd %in% list.dirs(working_directory)) {
dir.create(paste0(working_directory,cutoff_wd))}

stats_optimal_cutoff <- cutoff_stats[cutoff_stats$cutoff == optimal_cutoff, c("degree", "Probs")]
dd_plot_optimal = ggplot(stats_optimal_cutoff,aes(x=log(degree), y= log(Probs))) +
  geom_point() +
  geom_smooth(method="lm") +
  theme_bw() + 
  ggtitle(paste0("Chosen correlation cut-off [",optimal_cutoff, "]"))
print(dd_plot_optimal)
ggsave(filename = paste0("Degree_distribution_plot_", optimal_cutoff, ".pdf"), plot = dd_plot_optimal, 
       device = cairo_pdf, path = paste0(working_directory,cutoff_wd))


filt_cutoff_data = correlation_df_filt %>% dplyr::filter(rval > optimal_cutoff)
filt_cutoff_graph = igraph::graph_from_data_frame(filt_cutoff_data,directed=FALSE)
filt_cutoff_counts = ds[row.names(ds) %in% names(V(filt_cutoff_graph)),]
corresp_info = info_dataset[rownames(dd2)%in%rownames(info_dataset),]

print(paste("After using the optimal cutoff of",optimal_cutoff, "the number of edges =", 
            nrow(filt_cutoff_data), "and the number of nodes =", nrow(filt_cutoff_counts)))

heatmap_filtered_counts <- pheatmap::pheatmap(mat = filt_cutoff_counts,
                                              color=rev(RColorBrewer::brewer.pal(11, "RdBu")),
                                              scale="row",
                                              cluster_rows=T,
                                              cluster_cols=T,
                                              annotation_col=info_dataset[c(voi)],
                                              fontsize = 8,
                                              show_rownames = F, 
                                              show_colnames = F)

ggsave(filename = "Heatmap_topvar_genes.pdf", plot = heatmap_filtered_counts, device = cairo_pdf,
       path = paste0(working_directory,cutoff_wd))


```



## GFC calculation 

The group fold changes (GFCs) are calculated for each gene per condition. The defined variable of interest (voi) is used for the grouping of the samples.

```{r}

source(paste0(working_directory,"scripts/", "GFC_calculation.R" ))

# GFC_all_genes <- GFC_calculation(voi_id = voi)

GFC_all_genes_new_cluster <- GFC_calculation(voi_id = "new_cluster")



```



## Clustering

If cluster_algo = "auto"", five different cluster algorithms, namely cluster_label_prop, cluster_fast_greedy, cluster_louvain, cluster_infomap and cluster_walktrap are tested. Based on the highest modularity score, one of the algorithms is chosen for the actual clustering. Alternatively, one of the algorithms can be assigned to cluster_alog.  
I addition, the number of iterations can be specified (e.g. 100). If a gene is assigned to more clusters than specified in max_cluster_count_per_gene, it is painted white in the network, representing no definite cluster assignment. 


```{r}

source(paste0(working_directory,"scripts/", "cluster_calculation.R" ))



# 
# c("cluster_label_prop",
#                        "cluster_fast_greedy",
#                        "cluster_louvain",
#                        "cluster_infomap",
#                        "cluster_walktrap")

cluster_information <- cluster_calculation(igraph = filt_cutoff_graph,
                                           cluster_algo = "auto",
                                           no_of_iterations = 10,
                                           max_cluster_count_per_gene = 10,
                                           min_cluster_size = min_nodes_number_for_cluster,
                                           GFC = GFC_all_genes)

source(paste0(working_directory,"scripts/", "heatmap_clusters.R" ))

heatmap_cluster <- heatmap_clusters(data = cluster_information, cluster_cols = T)






# cluster_information_new_cluster <- cluster_calculation(igraph = filt_cutoff_graph,
#                                            cluster_algo = "auto",
#                                            no_of_iterations = 10,
#                                            max_cluster_count_per_gene = 10,
#                                            min_cluster_size = min_nodes_number_for_cluster,
#                                            GFC = GFC_all_genes_new_cluster)
# 
# source(paste0(working_directory,"scripts/", "heatmap_clusters.R" ))
# 
# heatmap_cluster_cluster <- heatmap_clusters(data = cluster_information_new_cluster, cluster_cols = T)
# 
# heatmap_insert_new_clustering(heatmap_data = cluster_table, height = height, compare_column_left = "merged",voi_id=voi_id,compare_column_right = "new_cluster", middle = "merged",color = "new_cluster")
# 




```


```{r message=FALSE, warning=FALSE}

source(paste0(working_directory,"scripts/", "clusters_samples_by_modules.R"))


norm_counts_fig3 <- read.delim("E:/RNA-Seq/Covid/Granulocytes_all/normalized counts.csv",header = T,row.names = 1,sep = ",",check.names = F)
norm_counts_fig3 <- norm_counts_fig3[,c(2:(ncol(norm_counts_fig3)-3))]
norm_counts_fig3 <- norm_counts_fig3[!duplicated(norm_counts_fig3[,ncol(norm_counts_fig3)]), ]
norm_counts_fig3 <- norm_counts_fig3[complete.cases(norm_counts_fig3), ]
row.names(norm_counts_fig3) <- norm_counts_fig3[,ncol(norm_counts_fig3)]
norm_counts_fig3 <- norm_counts_fig3[,c(-(ncol(norm_counts_fig3)))]


anno_files_fig3 <- read.delim("E:/RNA-Seq/Covid/Granulocytes_all/annotation.csv",header = T,sep = ",")


filt_count_mod <- gene_expression_over_modules(data_to_test = norm_counts_fig3,
                                               cluster_data=cluster_information_new_cluster,
                                               genes_of_int="all",
                                               order=NULL,
                                               anno_tab = anno_files_fig3,
                                               column_of_int = "X",
                                               top_n_modules = 11)


filt_count_mod <- merge(filt_count_mod,anno_files_fig3,by.x = "X", by.y="X")


filt_count_mod2 <- filt_count_mod %>%
  group_by(Module, Days_post_1st_symptoms, Donor,WHO_daily,severity) %>%
  summarise(value =mean(value))


filt_count_mod2 <- filt_count_mod2 %>%
  filter(Module %in% c("lightgreen","maroon","pink","darkgrey","darkgreen"))

#filt_count_mod2 <- filt_count_mod2 %>%
 # filter(!Module %in% c("lightgreen","maroon","pink","darkgrey","darkgreen"))


filt_count_mod2$Donor <- factor(filt_count_mod2$Donor,levels = c("771","763","764","777","782","759","761","758","742","746"))


ggplot(filt_count_mod2, aes(x = Days_post_1st_symptoms,  y = value, color = severity, fill = severity)) +
  #geom_line(size = 0.5,alpha=0.4)+
  scale_color_manual(values = c("#A49DC7","#522685"))+
  scale_fill_manual(values = c("#A49DC7","#522685"))+
    geom_smooth(method="auto")+
  theme_bw()+
  labs(title= "Cocena PAX modules",y="mean expression", x="Day after onset") + theme(panel.grid = element_line(inherit.blank = T),panel.grid.major = element_blank(),panel.grid.minor = element_blank(),panel.background = element_rect(fill = "white"),strip.background = element_rect(fill=c("grey")))+
  facet_grid(Module ~ ., scales = "free_y")


```



```{r message=FALSE, warning=FALSE}
#Granulos all reduced

source(paste0(working_directory,"scripts/", "modules_testing.R" ))

norm_counts_fig3 <- read.delim("E:/RNA-Seq/Covid/Granulocytes_all/normalized counts.csv",header = T,row.names = 1,sep = ",",check.names = F)
norm_counts_fig3 <- norm_counts_fig3[,c(2:(ncol(norm_counts_fig3)-3))]
norm_counts_fig3 <- norm_counts_fig3[!duplicated(norm_counts_fig3[,ncol(norm_counts_fig3)]), ]
norm_counts_fig3 <- norm_counts_fig3[complete.cases(norm_counts_fig3), ]
row.names(norm_counts_fig3) <- norm_counts_fig3[,ncol(norm_counts_fig3)]
norm_counts_fig3 <- norm_counts_fig3[,c(-(ncol(norm_counts_fig3)))]


anno_files_fig3 <- read.delim("E:/RNA-Seq/Covid/Granulocytes_all/annotation.csv",header = T,sep = ",")
anno_files_fig3$SampleID <- anno_files_fig3$ID
anno_files_fig3$ID_diagnosis <- paste0(anno_files_fig3$ID,"_",anno_files_fig3$Diagnosis)
anno_files_fig3$ID_severity <- paste0(anno_files_fig3$ID,"_",anno_files_fig3$severity)
anno_files_fig3$ID_severity_phase <- paste0(anno_files_fig3$ID,"_",anno_files_fig3$severity_phase,"_",anno_files_fig3$CoCena.G)

anno_files_fig3$ID_severity_phase_new <- paste0(anno_files_fig3$ID,"_",anno_files_fig3$severity_phase,"_",anno_files_fig3$CoCena.G,"_",anno_files_fig3$Donor,"_",anno_files_fig3$Days_post_1st_symptoms)

#anno_files_fig3$Neutro <- round(as.numeric(anno_files_fig3$Neutro),digits = 2)

anno_files_fig3$ID_severity_phase_neut <- paste0(anno_files_fig3$ID,"_",anno_files_fig3$severity_phase,"_",anno_files_fig3$Neutro)

# 
# anno_files_fig3_red <- subset(anno_files_fig3,!anno_files_fig3$ID %in% c("9456"))
# norm_counts_fig3_red <- subset(norm_counts_fig3,select = -c("9456"))
order_row <- c("indianred","maroon","darkorange","steelblue","gold","lightgreen","darkgreen","orchid","pink","darkgrey")


compare_external_signature(sample_file = norm_counts_fig3,# count file with symbols as rownames and samples as columns 
                             anno_file =  anno_files_fig3,# annotation with samples as rownames, needs column called
                             # "SampleID" !
                             grpvar = "ID_severity_phase_new",
                           cluster_info = cluster_information_new_cluster,# clsuter information data frame
                             range_GFC = 2,
                           clust_c = T,
                           order = NULL,
                           order_row = order_row)





```
